2.4/run-mongod.sh

```diff
@@ -14,108 +15,96 @@ export MONGODB_DATADIR=/var/lib/mongodb/data
+    mongo admin --eval "db.addUser({user: 'admin', pwd: '$MONGODB_ADMIN_PASSWORD', roles: [ 'dbAdminAnyDatabase', 'userAdminAnyDatabase' , 'readWriteAnyDatabase', 'clusterAdmin' ]});"
+  fi
 
-	# Create MongoDB database admin, if his password is specified with MONGODB_ADMIN_PASSWORD
-	if [ -v MONGODB_ADMIN_PASSWORD ]; then
-		echo "=> Creating an admin user with a ${MONGODB_ADMIN_PASSWORD} password in MongoDB"
-		mongo admin --eval "db.addUser({user: 'admin', pwd: '$MONGODB_ADMIN_PASSWORD', roles: [ 'dbAdminAnyDatabase', 'userAdminAnyDatabase' , 'readWriteAnyDatabase' ]});"
-	fi
+  # Create standard user with read/write permissions, in specified database ('production' by default).
+  mongo $MONGODB_DATABASE --eval "db.addUser({user: '${MONGODB_USERNAME}', pwd: '${MONGODB_PASSWORD}', roles: [ 'readWrite' ]});"
+  mongod -f $MONGODB_CONFIG_PATH --shutdown
 
-	# Create standard user with read/write permissions, in specified database ('production' by default).
-	mongo $MONGODB_DATABASE --eval "db.addUser({user: '${MONGODB_USERNAME}', pwd: '${MONGODB_PASSWORD}', roles: [ 'readWrite' ]});"
-	mongod -f $MONGODB_CONFIG_PATH --shutdown
+  # Create a empty file which indicates that the database users were created.
+  touch /var/lib/mongodb/data/.mongodb_users_created
 
-	# Create a empty file which indicates that the database users were created.
-	touch /var/lib/mongodb/data/.mongodb_users_created
+  # Check if the MongoDB daemon is down.
+  wait_for_mongo_down
+}
 
-	# Check if the MongoDB daemon is down.
-	down_test
+function cleanup() {
+  if [ ! -z "${MONGODB_REPLICA_NAME-}" ]; then
+    deregister && sleep 5
+  fi
+  echo "=> Shutting down MongoDB server ..."
+  kill -2 $(cat ${MONGODB_PID_FILE})
+  wait_for_mongo_down
+  exit 0
 }
 
+cache_container_addr
+
 # Generate config file for MongoDB
 envsubst < ${MONGODB_CONFIG_PATH}.template > $MONGODB_CONFIG_PATH
 
 if [ "$1" = "mongod" ]; then
-
-	if ! [[ -v MONGODB_USERNAME && -v MONGODB_PASSWORD && -v MONGODB_DATABASE ]]; then
-		usage
-	fi
-
-	if [ ! -f /var/lib/mongodb/data/.mongodb_users_created ]; then
-		# Create default MongoDB user and administrator.
-		create_mongodb_users
-	fi
-
-	unset_env_vars
-
-	# Start MongoDB service with enabled database authentication.
-	exec mongod -f $MONGODB_CONFIG_PATH --auth
+  if ! [[ -v MONGODB_USERNAME && -v MONGODB_PASSWORD && -v MONGODB_DATABASE ]]; then
+    usage
+  fi
+
+  if [ ! -f /var/lib/mongodb/data/.mongodb_users_created ]; then
+    # Create default MongoDB user and administrator.
+    create_mongodb_users
+  fi
+
+  unset_env_vars
+
+  mongo_cmd_args="-f $MONGODB_CONFIG_PATH --noprealloc --smallfiles --oplogSize 64"
+
+  if [ ! -z "${MONGODB_REPLICA_NAME-}" ]; then
+    mongo_cmd_args+=" --replSet ${MONGODB_REPLICA_NAME}"
+    ( mkfifo /tmp/replica_supervisor_log && cat /tmp/replica_supervisor_log ) &
+    /var/lib/mongodb/replica_supervisor.sh &
+    trap 'cleanup' SIGINT SIGTERM
+    mongod $mongo_cmd_args & mongo_pid=$!
+    wait $mongo_pid
```

Comment:
yeah, but that is enough, usually the deregister takes less than second (it just sent the request to all replicas)